### PR TITLE
Provide autocomplete suggestions in SearchRouterList

### PR DESCRIPTION
**Pull request by SzymczakJ**
### Details

This PR adds autocomplete in SearchRouter for simple keys like categories, tags, etc.. Searching with 'in:' keyword is not yet fully supported(reportIDs instead of full names will be visible in text input)

### Fixed Issues

$ https://github.com/Expensify/App/issues/50942
$ https://github.com/Expensify/App/issues/50947
$ https://github.com/Expensify/App/issues/50957
$ https://github.com/Expensify/App/issues/50952
$ https://github.com/Expensify/App/issues/50953
$ https://github.com/Expensify/App/issues/50954
$ https://github.com/Expensify/App/issues/50955
$ https://github.com/Expensify/App/issues/50956
$ https://github.com/Expensify/App/issues/50948
$ https://github.com/Expensify/App/issues/50945
PROPOSAL:

### Tests

1. Open SearchRouter and test these autocomplete keys: category, tag, taxRate, type, expenseType, card, from, to, status,  and currency.
2. Verify that when category, tag or currency keys are working the way [this comment](https://github.com/Expensify/App/pull/51237#issuecomment-2430179534) specified it.
3. Verify that after clearing input or single autocomplete key the autocomplete suggestions are cleared.
4. Verify that after clicking on autocomplete suggestion it is added correctly to text input.

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/9df88b21-d26f-44e3-94f9-91827a930f7a

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/d8f6d256-c08b-470f-b743-a3d78f388961

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/81c1a2c7-7135-42c2-9097-18982e80c36c

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/3940ed06-b698-456a-95d7-22bcfde089a1

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/430ee71e-c4a1-4088-b4c2-d19b3c055fee

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/4bd54f7e-4518-494e-8065-0132a8a99450

</details>

### CODE REVIEW

**Comment by SzymczakJ** [comment]
Should we somehow limit the list of autosuggestions @luacmartins? This is POC implementation of autosuggestions for some of filters(the ones linked in description) 

https://github.com/user-attachments/assets/46bbf9ba-b956-4ffa-9f5e-0adeb9cb6a3c

I don't see much designs connected to autocomplete.

**Comment by luacmartins** [comment]
@SzymczakJ I think we should only offer suggestions once the user typed at least one character after the filter key, e.g. `currency:A`. tagging @trjExpensify @JmillsExpensify for your thoughts as well.

**Comment by JmillsExpensify** [comment]
Agree with Carlos that we'd suggest on the first character after the filter key.

**Comment by luacmartins** [comment]
@SzymczakJ we discussed internally and the behavior we want for this is:

1. `category:` would show a list of 10 "recently used" across all policies, from the `RecentlyUsedTags`, etc Onyx key. This can be an empty list.
2. `category:a` would filter results from all `policyCategories` for categories starting with `a` and show those suggestions.

Does that make sense?

File: src/libs/SearchAutocompleteUtils.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+function parseForAutocomplete(text: string) {
+    try {
+        const parsedAutocomplete = autocompleteParser.parse(text) as SearchAutocompleteResult;
+        return parsedAutocomplete;
+    } catch (e) {
+        console.error(`Error when parsing autocopmlete}"`, e);
+    }
+}
+
+function getAutoCompleteTagsList(allPoliciesTagsLists: OnyxCollection<PolicyTagLists>, policyID?: string) {
```
Comment (Kicu) [inline src/libs/SearchAutocompleteUtils.ts:17]:
```suggestion
function getAutoCompleteTagList(allPoliciesTagsLists: OnyxCollection<PolicyTagLists>, policyID?: string) {
```

File: src/libs/SearchAutocompleteUtils.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+                return getTagNamesFromTagsLists(policyTagLists);
+            })
+            .flat()
+            .forEach((tag) => uniqueTagNames.add(tag));
+        return Array.from(uniqueTagNames);
+    }
+    return getTagNamesFromTagsLists(singlePolicyTagsList);
+}
+
+function getAutocompleteCategoriesList(allPolicyCategories: OnyxCollection<PolicyCategories>, policyID?: string) {
```
Comment (Kicu) [inline src/libs/SearchAutocompleteUtils.ts:33]:
```suggestion
function getAutocompleteCategoryList(allPolicyCategories: OnyxCollection<PolicyCategories>, policyID?: string) {
```

File: src/libs/SearchAutocompleteUtils.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    } catch (e) {
+        console.error(`Error when parsing autocopmlete}"`, e);
+    }
+}
+
+function getAutoCompleteTagsList(allPoliciesTagsLists: OnyxCollection<PolicyTagLists>, policyID?: string) {
+    const singlePolicyTagsList: PolicyTagLists | undefined = allPoliciesTagsLists?.[`${ONYXKEYS.COLLECTION.POLICY_TAGS}${policyID}`];
+    if (!singlePolicyTagsList) {
+        const uniqueTagNames = new Set<string>();
+        const tagListsUnpacked = Object.values(allPoliciesTagsLists ?? {}).filter((item) => !!item) as PolicyTagLists[];
```
Comment (Kicu) [inline src/libs/SearchAutocompleteUtils.ts:21]: is possible to use default argument instead, and do
```ts
getAutoCompleteTagsList(allPoliciesTagsLists: OnyxCollection<PolicyTagLists> = {}, policyID?: string)
```
?

File: src/libs/SearchAutocompleteUtils.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        tagListsUnpacked
+            .map((policyTagLists) => {
+                return getTagNamesFromTagsLists(policyTagLists);
```
Comment (Kicu) [inline src/libs/SearchAutocompleteUtils.ts:24]:
```suggestion
            .map(getTagNamesFromTagsLists)
```

Comment (Kicu) [reply src/libs/SearchAutocompleteUtils.ts:24]: I believe the extra wrapper arrow function is redundant

File: src/components/Search/SearchRouter/SearchRouter.tsx [hunk 5]
```diff
@@ -39,18 +44,40 @@ function SearchRouter({onRouterClose}: SearchRouterProps) {
...
+    const [currencyList] = useOnyx(ONYXKEYS.CURRENCY_LIST);
+    const currencyAutocompleteList = Object.keys(currencyList ?? {});
+    const [allPoliciesTagsLists] = useOnyx(ONYXKEYS.COLLECTION.POLICY_TAGS);
+    const tagAutocompleteList = useMemo(() => getAutoCompleteTagsList(allPoliciesTagsLists, activeWorkspaceID), [allPoliciesTagsLists, activeWorkspaceID]);
+    const allTaxRates = getAllTaxRates();
+    const taxAutocompleteList = useMemo(() => getAutocompleteTaxList(allTaxRates, policy), [policy, allTaxRates]);
+    const [cardList = {}] = useOnyx(ONYXKEYS.CARD_LIST);
+    const cardsAutocompleteList = Object.values(cardList ?? {}).map((card) => card.bank);
+    const personalDetails = usePersonalDetails();
+    const participantsAutocompleteList = Object.values(personalDetails)
```
Comment (Kicu) [inline src/components/Search/SearchRouter/SearchRouter.tsx:76]: Good job, but still:
image



😆 but seriously NAB and if we need all this data then we need it

**Review by Kicu — COMMENTED** [review]
On a quick glance the logic and data flow looks good 👌 

I hope that in future we can somehow split the logic so that `SearchRouter` does not become a "god component"

File: src/libs/SearchAutocompleteUtils.ts (hunk 3 above)
Comment (SzymczakJ) [reply src/libs/SearchAutocompleteUtils.ts:21]: I believe our hands are tied by OnyxCollection which can be undefined.

File: src/components/Search/SearchRouter/SearchRouter.tsx (hunk 5 above)
Comment (luacmartins) [reply src/components/Search/SearchRouter/SearchRouter.tsx:77]: Yea, I'm not sure there's a way around it. Maybe we can create a custom hook, useAutocomplete so the logic lives outside the component?

**Review by luacmartins — COMMENTED** [review]
I know this is still WIP so maybe you might be aware of these issues already, but I wanted to point them out anyways so we can get ahead and make sure they don't make to the final version of the PR:

1. I think we should limit the number of suggestions to maybe 10 sorted alphabetically. You can see on the list that if I type from, a full list of all personalDetails comes up, which seems a bit useless. 
2. Cycling through the options should update the query string, e.g. when I press the arrow keys to select a different option, the query in the router should update to include the currently active suggestion.
3. Selecting an autocomplete suggestion runs the query, but it drops the filter key. Note in the video how I selected `from:t3@t.com` but the final search dropped the `from:` part

https://github.com/user-attachments/assets/149b8c80-9b17-47bf-8e6a-e9e7d9e8bad7

File: src/components/Search/SearchRouter/SearchRouter.tsx [hunk 6]
```diff
@@ -39,18 +48,53 @@ function SearchRouter({onRouterClose}: SearchRouterProps) {
...
     const contextualReportID = useNavigationState<Record<string, {reportID: string}>, string | undefined>((state) => {
         return state?.routes.at(-1)?.params?.reportID;
     });
+
+    const activeWorkspaceID = useActiveWorkspaceFromNavigationState();
+    const policy = usePolicy(activeWorkspaceID);
+    const typesAutocompleteList = Object.values(CONST.SEARCH.DATA_TYPES);
+    const statusesAutocompleteList = Object.values({...CONST.SEARCH.STATUS.TRIP, ...CONST.SEARCH.STATUS.INVOICE, ...CONST.SEARCH.STATUS.CHAT, ...CONST.SEARCH.STATUS.TRIP});
+    const expenseTypes = Object.values(CONST.SEARCH.TRANSACTION_TYPE);
+    const allTaxRates = getAllTaxRates();
```
Comment (Kicu) [inline src/components/Search/SearchRouter/SearchRouter.tsx:68]: this is a duplicate of line 56:
`    const taxRates = getAllTaxRates();`

remove duplica te and please call it simply `taxRates`. I find the name `allXXXX` confusing

File: src/components/Search/SearchRouter/SearchRouter.tsx [hunk 7]
```diff
@@ -39,18 +48,53 @@ function SearchRouter({onRouterClose}: SearchRouterProps) {
...
     const [textInputValue, debouncedInputValue, setTextInputValue] = useDebouncedState('', 500);
-    const [userSearchQuery, setUserSearchQuery] = useState<SearchQueryJSON | undefined>(undefined);
     const contextualReportID = useNavigationState<Record<string, {reportID: string}>, string | undefined>((state) => {
         return state?.routes.at(-1)?.params?.reportID;
     });
+
+    const activeWorkspaceID = useActiveWorkspaceFromNavigationState();
+    const policy = usePolicy(activeWorkspaceID);
+    const typesAutocompleteList = Object.values(CONST.SEARCH.DATA_TYPES);
+    const statusesAutocompleteList = Object.values({...CONST.SEARCH.STATUS.TRIP, ...CONST.SEARCH.STATUS.INVOICE, ...CONST.SEARCH.STATUS.CHAT, ...CONST.SEARCH.STATUS.TRIP});
```
Comment (Kicu) [inline src/components/Search/SearchRouter/SearchRouter.tsx:66]:
```suggestion
    const statusAutocompleteList = Object.values({...CONST.SEARCH.STATUS.TRIP, ...CONST.SEARCH.STATUS.INVOICE, ...CONST.SEARCH.STATUS.CHAT, ...CONST.SEARCH.STATUS.TRIP});
```

**Review by Kicu — APPROVED** [review]
very small stuff, feel free to open PR when you're ready.

**Comment by luacmartins** [comment]
Both @rayane-djouah and @ikevin127 have been involved and reviewing Search PRs for a while, so they both have context and should be able to handle these PRs.

**Review by luacmartins — CHANGES_REQUESTED** [review]
Nice, this is looking really good. Just gonna report a few bugs that I found

**Comment by luacmartins** [comment]
BUG?: Pressing tab cycles through the options as it should, but it also removes the focus from the text input so the user has to click back into the input. This doesn't happen with the arrow navigation so users can keep typing into the input. We should make this behavior consistent.

**Tab:**
https://github.com/user-attachments/assets/969dad70-5d93-46b8-a903-8959ed3a9326

**Arrows**

https://github.com/user-attachments/assets/c559ac2c-6b8e-437d-86b1-6230e1337056

**Comment by luacmartins** [comment]
BUG: We should make suggestions case insensitive. For example, when searching for a currency, it correctly returns the results if I search in upper case, but nothing when searching in lower case 

https://github.com/user-attachments/assets/8dbd18bd-baa8-41f2-9897-ced0be5867e5

**Comment by luacmartins** [comment]
BUG: Searching for categories with space results in no results and incorrect final syntax.

https://github.com/user-attachments/assets/dc6f7884-9032-4ed4-a8e5-80c3ac70740f

**Comment by SzymczakJ** [comment]
> BUG?: Pressing tab cycles through the options as it should, but it also removes the focus from the text input

This was also the tab behaviour of ChatFinderPage. I tried to make it not defocus TextInput but after investigating it for some time, I think we would need to modify https://github.com/Expensify/react-native-key-command library, which makes this "bug" harder to fix.
I think this is a good candidate for Search 2.5 polish, or maybe, let's decide that's not a bug 😄 ? As I said before ChatFinderPage had identical behaviour. WDYT @luacmartins?

**Comment by SzymczakJ** [comment]
@ikevin127 kind bump 🙇

**Comment by luacmartins** [comment]
>This was also the tab behaviour of ChatFinderPage. I tried to make it not defocus TextInput but after investigating it for some time, I think we would need to modify https://github.com/Expensify/react-native-key-command library, which makes this "bug" harder to fix.

I think it's ok to do this as a polish or wait until we get feedback that we should implement this.

**Comment by ikevin127** [comment]
I'll start reviewing in ~ 1 hour.

**Comment by ikevin127** [comment]
## Reviewer Checklist

- [ ] I have verified the author checklist is complete (all boxes are checked off).
- [ ] I verified the correct issue is linked in the `### Fixed Issues` section above
- [ ] I verified testing steps are clear and they cover the changes made in this PR
    - [ ] I verified the steps for local testing are in the `Tests` section
    - [ ] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [ ] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [ ] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [ ] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [ ] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [ ] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [ ] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [ ] I verified that all the inputs inside a form are aligned with each other.
    - [ ] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

File: src/libs/SearchAutocompleteUtils.ts [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+    }
+    return getTagNamesFromTagsLists(singlePolicyTagsList);
+}
+
+function getAutocompleteRecentTags(allRecentTags: OnyxCollection<RecentlyUsedTags>, policyID?: string) {
+    const singlePolicyRecentTags: RecentlyUsedTags | undefined = allRecentTags?.[`${ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_TAGS}${policyID}`];
+    if (!singlePolicyRecentTags) {
+        const uniqueTagNames = new Set<string>();
+        Object.values(allRecentTags ?? {})
+            .map((testVar) => Object.values(testVar ?? {}))
```
Comment (luacmartins) [inline src/libs/SearchAutocompleteUtils.ts:36]:
```suggestion
            .map((recentTag) => Object.values(recentTag ?? {}))
```

File: src/libs/SearchAutocompleteUtils.ts [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+function getAutocompleteRecentCategories(allRecentCategories: OnyxCollection<RecentlyUsedCategories>, policyID?: string) {
+    const singlePolicyRecentCategories = allRecentCategories?.[`${ONYXKEYS.COLLECTION.POLICY_RECENTLY_USED_CATEGORIES}${policyID}`];
+    if (!singlePolicyRecentCategories) {
+        const uniqueCategoryNames = new Set<string>();
+        Object.values(allRecentCategories ?? {}).map((policyCategories) => Object.values(policyCategories ?? {}).forEach((category) => uniqueCategoryNames.add(category)));
+        return Array.from(uniqueCategoryNames);
+    }
+    return Object.values(singlePolicyRecentCategories ?? {}).map((category) => category);
+}
```
Comment (luacmartins) [inline src/libs/SearchAutocompleteUtils.ts:62]: NAB maybe we could DRY the logic in some of these functions too since the logic is 95% identical

**Review by luacmartins — APPROVED** [review]
LGTM. Left a few small comments. We also have conflicts now and product is still testing this branch to make sure things are what we expect.

**Review by luacmartins — CHANGES_REQUESTED** [review]
Let's address the variable reaming/typos above

**Comment by ikevin127** [comment]
## ⚠️ **Found two issues while reviewing:**

1. **Multiple options for 1 autocomplete key results in infinite loading**

> [!note]
> ### Explanation
> When selecting 2 or more options and performing search we're stuck with infinite loading for the following keys: **type**, **taxRate** and **status**.
>
> Below video demonstrates the behaviour for the **type** key.

https://github.com/user-attachments/assets/defa47b7-67a6-4013-ac2b-949a530b89b7

2. **[UX] Selecting 1 autocomplete option adds [space] then when comma is added after space, no autocomplete options are shown**

> [!note]
> ### Explanation
> The issue here is that when selecting 1 autocomplete option, [space] is added in the input after the option which if we type , (comma) after the space, other autocomplete options don't show up. The only way options continue to show up is if we delete the space that is added automatically, then type , (comma) right after the first selected option.
>
> Below video demonstrates the behaviour.

https://github.com/user-attachments/assets/b7b04207-11fd-46ce-833e-c23d0ea78702


##

cc @SzymczakJ @luacmartins 

I'll move forward with the reviewer checklist in the meantime, and will approve once the above issues are addressed and conflicts resolved.

**Comment by Kicu** [comment]
@ikevin127 @luacmartins @SzymczakJ
In order to speed implementing things up, I'm already branched from this PR and working on the next steps of autocomplete.  Mainly the `accountID/reportID` <---> `user login/chatname` substitution, so that the query the user sees is always correct.

Because of that I would like to suggest any NAB's and other improvements are not done here, but I will instead do them in my PR 😁 it will greatly help me.
Let's fix only critical and problematic bugs here.

File: src/components/Search/SearchRouter/SearchRouter.tsx [hunk 10]
```diff
@@ -155,64 +197,186 @@ function SearchRouter({onRouterClose}: SearchRouterProps) {
...
+                        .slice(0, 10);
+                    setAutocompleteSuggestions(
+                        filteredCards.map((card) => ({
+                            text: `${CONST.SEARCH.SYNTAX_FILTER_KEYS.CARD_ID}:${card}`,
+                            query: `${card}`,
+                        })),
+                    );
+                    return;
+                }
+                default: {
```
[the remark this answers is not in the document]
Comment (SzymczakJ) [reply src/components/Search/SearchRouter/SearchRouter.tsx:274]: I know this code looks easily "DRYable", but with Mateusz [PR](https://github.com/Expensify/App/pull/51633) these switch cases will become much more distinct. Let's not add conflicts to his branch, as he said in https://github.com/Expensify/App/pull/51237#issuecomment-2443561880 😄. After merging these two PRs we could put up a small clean-up PR.

File: src/libs/SearchAutocompleteUtils.ts (hunk 9 above)
Comment (SzymczakJ) [reply src/libs/SearchAutocompleteUtils.ts:62]: Same as in https://github.com/Expensify/App/pull/51237#discussion_r1819798963

**Comment by SzymczakJ** [comment]
> 1. Multiple options for 1 autocomplete key results in infinite loading

It results in infinite loading, because it doesn't make sense from UI point of view(we cannot be on an Expense tab and on Chat tab on the same moment). On the other hand I think that we shouldn't limit the user, even though his query may produce something which doesn't make sense. Maybe we could provide some modal on SearchResultsPage that tell user that 'This query doesn't make sense'. WDYT @luacmartins 

> 2. [UX] Selecting 1 autocomplete option adds [space] then when comma is added after space, no autocomplete options are shown

No autocomplete was added for input **"category:X ,"**  because of the limitations of our parser. Here is my solution for this: when user adds "," to add another autocomplete then the spaces before "," are trimmed so our parser can parse it properly. Please check it out @luacmartins @ikevin127 

https://github.com/user-attachments/assets/7bcd7ea4-04a4-44ec-9a21-4f80ada7df49

**Review by ikevin127 — APPROVED** [review]
Looks great, thanks to everybody for addressing and the issues! :shipit:

**Comment by luacmartins** [comment]
NAB I think we can get smarter with the suggested statuses and only suggest statuses that apply to the current type
Screenshot 2024-10-29 at 12 21 47 PM

**Review by luacmartins — APPROVED** [review]
LGTM and tests well. Just waiting on final approval from the product team!

File: src/components/Search/SearchRouter/SearchRouter.tsx [hunk 11]
```diff
@@ -101,64 +143,190 @@ function SearchRouter({onRouterClose}: SearchRouterProps) {
...
+                        .sort()
+                        .slice(0, 10);
+                    setAutocompleteSuggestions(
+                        filteredCards.map((card) => ({
+                            text: `${CONST.SEARCH.SYNTAX_FILTER_KEYS.CARD_ID}:${card}`,
+                            query: `${card}`,
+                        })),
+                    );
+                    return;
+                }
```
Comment (s77rt) [inline src/components/Search/SearchRouter/SearchRouter.tsx:273]: We should have used the card last four as well for auto suggestion. Coming from https://github.com/Expensify/App/issues/57994

**Pull request merged**